sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala

```diff
@@ -183,6 +183,13 @@ case class InMemoryTableScanExec(
   private val stats = relation.partitionStatistics
   private def statsFor(a: Attribute) = stats.forAttribute(a)
 
+  // For some ColumnStats, for instance, ObjectColumnStats always has nulls for lower and upper
+  // bounds.
+  private def nullSafeEval(
+      attr: AttributeReference)(func: AttributeReference => Expression): Expression = {
+    attr.isNull || func(attr)
```

Comment:
this adds extra runtime null check and may introduce perf regression. How about we follow the hive partition pruning and only create filters for non-complex type? e.g.
```
    object ExtractableLiteral {
      def unapply(expr: Expression): Option[Expression] = {
        if (expr.dataType.isInstanceOf[AtomicType]) Some(expr) else None
      }
    }
...
case EqualTo(a: AttributeReference, ExtractableLiteral(l)) =>
```

Reply:
Thanks, @cloud-fan.

Reply:
this basically means turning off the filtering for complex types. Despite this may be not a big deal, as probably we won't have complex types often here, can't we instead add the isNull filter only for complex types?

Reply:
hmmmmm .. if this can whitelist the cases we support, I thought it's okay to use the suggestion above. BTW, looks we should exclude binary type too. It will still supports itnull or isnotnull though.